packages/nextjs/src/middlewareClient.ts

```diff
@@ -102,3 +114,228 @@ export function createMiddlewareClient<
+ * export function middleware(request: NextRequest) {
+ *   const [supabaseClient, nextResponse] = createMiddlewareClient2(request, {})
+ *
+ *   const { data: { user } } = await supabaseClient.auth.getUser()
+ *
+ *   if (!user) {
+ *     // there's no user session, ask them to log in
+ *     return NextResponse.redirect(new URL('/sign-in-page-in-your-app'))
+ *   }
+ *
+ *   return nextResponse()
+ * }
+ * ```
+ */
+export function createMiddlewareClient2<
+	Database = any,
+	SchemaName extends string & keyof Database = 'public' extends keyof Database
+		? 'public'
+		: string & keyof Database,
+	Schema extends GenericSchema = Database[SchemaName] extends GenericSchema
+		? Database[SchemaName]
+		: any
+>(
+	req: NextRequest,
+	{
+		supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL,
+		supabaseKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY,
+		options,
+		cookieOptions
+	}: {
+		supabaseUrl?: string;
+		supabaseKey?: string;
+		options?: SupabaseClientOptionsWithoutAuth<SchemaName>;
+		cookieOptions?: CookieOptionsWithName;
+	} = {}
+): [SupabaseClient<Database, SchemaName, Schema>, () => NextResponse] {
+	if (!supabaseUrl || !supabaseKey) {
+		throw new Error(
+			'either NEXT_PUBLIC_SUPABASE_URL and NEXT_PUBLIC_SUPABASE_ANON_KEY env variables or supabaseUrl and supabaseKey are required!'
+		);
+	}
+
+	const storage = new RequestCookiesAuthStorageAdapter(req);
+
+	const client = createSupabaseClient<Database, SchemaName, Schema>(supabaseUrl, supabaseKey, {
+		...options,
+		global: {
+			...options?.global,
+			headers: {
+				...options?.global?.headers,
+				'X-Client-Info': `${PACKAGE_NAME}@${PACKAGE_VERSION}`
+			}
+		},
+		auth: {
+			storage
+		}
+	});
+
+	const responseFn: () => NextResponse = () => {
+		const requestCookies = (req.headers.get('cookie') || '')
+			.split(/\s*;\s*/g)
+			.map((part) => part.split(/=/g))
+			.reduce((a, [name, value]) => {
+				a[name] = value;
+				return a;
+			}, {} as { [cookie: string]: string });
+
+		const removeKeys = {
+			...storage.deleteItems,
+			...storage.items
+		};
+
+		// remove the cookies and their chunks that the client decided
+		// needed removing or setting
+		Object.keys(removeKeys).forEach((key) => {
+			delete requestCookies[key];
+
+			const chunkNames = Object.keys(requestCookies).filter(
+				(k) => k.startsWith(key) && k.length > key.length + 1 && k[key.length] === '.'
+			);
+
+			chunkNames.forEach((chunk) => {
+				delete requestCookies[chunk];
+			});
+		});
+
+		// after the cookies and their chunks have been cleaned up, we
+		// can assign the set items directly; no chunking is needed as
+		// these values will be passed down to the NextJS component /
+		// page / route
+		Object.keys(storage.items).forEach((key) => {
+			requestCookies[key] = storage.items[key];
+		});
+
+		// reconstitute the cookie header on the request
+		req.headers.set(
+			'cookie',
+			Object.entries(requestCookies)
+				.map(([key, value]) => `${key}=${value}`)
+				.join('; ')
+		);
```

Comment:
ah so is this where we make sure that the request contains the newly refreshed session so subsequent calls to `getUser` / `getSession` in the server-side components?